flesh out core repo and add user_management app.

jimabramson opened the pull request:
~~Establishing a repo for reusable (domain-neutral) django apps that are useful with open edx (but do not depend on it).~~

EDIT: I've been talked out of an omnibus repo full of configuration tools.  Instead this is repositioned as a place for automation-friendly management commands.  It's a separate repo b/c it can be used with any Django app.

@clintonb 
@e0d - need feedback on management commands, welcome feedback on the rest
@cpennington

Commit:
Flesh out core repo and add user and group management.

cpennington commented on .travis.yml:
```diff
@@ -0,0 +1,22 @@
+sudo: false
+
+# Cache the pip directory. "cache: pip" doesn't work due to install override. See https://github.com/travis-ci/travis-ci/issues/3239.
+cache:
```
Did you see the `cache: custom_install` option that they added?

jimabramson replied:
good call, thanks, will add that in.

cpennington commented:
You have a number of `djangoo-e-extensions` in these changes? Typo?

cpennington commented on docs/features.rst:
```diff
@@ -0,0 +1,28 @@
+Feature Toggling
+================
```
I don't think we want this in this repo.

cpennington commented on docs/getting_started.rst:
```diff
@@ -0,0 +1,86 @@
++-----------------------------------------------------+----------------------------------------------------------------------------+--------------------------------------------------------------------------+
+| Setting                                             | Description                                                                | Value                                                                    |
++=====================================================+============================================================================+==========================================================================+
+| SOCIAL_AUTH_EDX_OIDC_KEY                            | OAuth 2.0 client key                                                       | (This should be set to the value generated when the client was created.) |
```
No need for social auth in these, right?

mattdrayer commented:
Just wondering why we decided to create an aggregate projects repo versus individual repos for each project?

mattdrayer commented:
Also is the idea of this repo to serve as a place to move code from the "common" and "openedx" areas of edx-platform?

jimabramson commented:
@cpennington thanks for the catches, and sorry about the messy stuff - byproducts of a silly mistake with cookiecutter.  I've fixed up the issues you called out.

@mattdrayer this is meant to be home to multiple django apps (for example, config_models in platform is its own django app, and I'm specifically planning to add that in here).  The idea of this repo is not to subsume common/openedx, but only parts that would conceivably be reused in any arbitrary django app (but stuff particularly useful at edx).  The point of departure is the django-extensions project (not only in name).

Based on a very quick look around, here are other examples of things that live in platform that might fit in here:
* common/djangoapps/geoinfo
* common/djangoapps/monitoring
* common/djangoapps/performance
* common/djangoapps/request_cache

mattdrayer commented:
When I see the name "edx-django-extensions" I immediately picture a collection of utilities that edX has constructed which extend the Django web application framework.  This doesn't seem to be the case, however.  This seems like more of an application services project?

jimabramson commented:
@e0d @feanil FYI - I have added a brief doc page for the management commands which we discussed in the context of https://openedx.atlassian.net/browse/DEVOPS-3339 - this is primary motivation for this PR so your review/signoff on that (and the code in user_management) is requested.

jimabramson commented:
@mattdrayer I'm actually totally open to, and cool with, renaming the repo/project if another title seems better.  (Though @arbabnazar might get annoyed with my recurring admin requests on this repo :-))

The main goal is to have a place for our reusable utility django apps so we can use them in more than one place.  I don't know if that's the same as "application services"

mattdrayer commented:
I have to admit, I feel like this project is just going to become another junk-drawer of random apps and utilities.  Wouldn't it be better to have a separate repo for each package?

jimabramson commented:
@mattdrayer I really don't feel that strongly about it; with minor changes (just bump up a package by one level, and rename the repo) it can be just that.  At the same time, I'd like to understand better, why is it better to have multiple repos?  This just seems like slightly more admin and doc overhead and more separate requirements to manage in the projects that will include them.  To go with the junk drawer metaphor, the alternative strikes me as dumping the drawer out on the table.

mattdrayer commented:
I just wonder if all projects will require all extensions, and what happens if we want to upgrade only a single project -- you know all the regular code management issues :)  I think I'm personally scarred by this sort of pattern due to a past life where I was responsible for a thing like this, which seemed like a good idea at the beginning but quickly became a burden.

jimabramson commented:
OK I'm going to go ahead and modify this to be a single-app repo, and will put in a request to have it renamed: thinking edx-user-management (it's already a mouthful without much room to fit 'django' in there)

mattdrayer commented on README.rst:
```diff
@@ -1 +1,59 @@
+
+__ http://code.edx.org/
+
+EdX Django Extensions  |Travis|_ |Codecov|_
```
You'll need to rename this to "edX User Management"

mattdrayer commented on README.rst:
```diff
@@ -1 +1,59 @@
+Documentation
+-------------
+
+The docs for EdX Django Extensions are on Read the Docs:  https://edx-django-extensions.readthedocs.org.
```
"edX User Management"

mattdrayer commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+# -*- coding: utf-8 -*-
+#
+# EdX Django Extensions documentation build configuration file, created by
```
edX User Management

mattdrayer commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+master_doc = 'index'
+
+# General information about the project.
+project = u'EdX Django Extensions'
```
edX User Management

mattdrayer commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+# html_file_suffix = None
+
+# Output file base name for HTML help builder.
+htmlhelp_basename = 'edx-django-extensionsdoc'
```
edx-user-managementdoc

mattdrayer commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+# Grouping the document tree into LaTeX files. List of tuples
+# (source start file, target name, title, author, documentclass [howto/manual]).
+latex_documents = [
+    ('index', 'edx-django-extensions.tex', u'EdX Django Extensions Documentation',
```
same here

mattdrayer commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+# One entry per manual page. List of tuples
+# (source start file, name, description, authors, manual section).
+man_pages = [
+    ('index', 'EdX Django Extensions', u'EdX Django Extensions Documentation',
```
rename here as well

mattdrayer commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+# (source start file, target name, title, author,
+#  dir menu entry, description, category)
+texinfo_documents = [
+    ('index', 'EdX Django Extensions', u'EdX Django Extensions Documentation',
```
rename

mattdrayer commented on docs/index.rst:
```diff
@@ -0,0 +1,17 @@
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+EdX Django Extensions
```
edX User Management

mattdrayer commented on setup.py:
```diff
@@ -0,0 +1,31 @@
+from setuptools import setup, find_packages
+
+setup(
+    name='edx-django-extensions',
```
edx-user-management

mattdrayer commented on setup.py:
```diff
@@ -0,0 +1,31 @@
+setup(
+    name='edx-django-extensions',
+    version='0.0.1',
+    description='Reusable automation and configuration utilities for edX Django applications',
```
update description

mattdrayer commented on setup.py:
```diff
@@ -0,0 +1,31 @@
+    version='0.0.1',
+    description='Reusable automation and configuration utilities for edX Django applications',
+    author='edX',
+    url='https://github.com/edx/edx-django-extensions',
```
edx-user-management

jimabramson commented:
@mattdrayer thanks - i will go ahead and update the titles.  After my last comment I decided that a better framing for the pared-down form of this "edx management commands" and used that for renaming the app/package.  That is how I'll rename in the places you've pointed out.

@cpennington @e0d any concerns you want to inject here?

jimabramson commented:
@feanil this

jimabramson commented:
@feanil @fredsmith @adampalay this is the PR mentioned this morning, with mgmt commands for creating users/groups/group assignments.  The repo name needs to change but I'll tend to that after I'm sure we agree on what it should be changed to :-)

vast majority of the PR is boilerplate.   interesting part is the two commands (and their tests) in edx_management_commands/management_commands/

e0d commented on README.rst:
```diff
@@ -1 +1,59 @@
+
+Please do not report security issues in public. Please email security@edx.org.
+
+Mailing List and IRC Channel
```
Slack?

e0d commented on docs/commands.rst:
```diff
@@ -0,0 +1,73 @@
+
+.. code-block:: bash
+
+    $ python manage.py put_group privileged_group -p myapp:Thing:add_thing myapp:Thing:delete_thing
```
Is there any mechanism for getting the available groups?  Seems useful for debugging.

jimabramson replied:
no, there isn't.  but agree it could be useful to add.

e0d commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+
+# General information about the project.
+project = u'EdX Management Commands'
+copyright = u'2015, edX'
```
2016

e0d commented on edx_management_commands/management_commands/management/commands/put_group.py:
```diff
@@ -0,0 +1,118 @@
+            new_permissions.append(new_permission)
+
+        add_permissions = [p for p in new_permissions if p not in old_permissions]
+        remove_permissions = [p for p in old_permissions if p not in new_permissions]
```
Can we add help text to the web ui that discusses the dangers of manual changes?  Other choice might be to fail if len(remove_permissions) > 0 and have a --force-removal option.

jimabramson replied:
That could be any web UI (these are just management commands which should work with any Django project/app).  If edx-platform is the particular concern, we can make a separate PR for that (but would like more detail on what text seems useful to expose and where)

e0d commented on edx_management_commands/management_commands/management/commands/put_user.py:
```diff
@@ -0,0 +1,105 @@
+    def add_arguments(self, parser):
+        parser.add_argument('username')
+        parser.add_argument('email')
+        parser.add_argument('--superuser', dest='is_superuser', action='store_true')
```
Maybe crazy, but I suppose it's too over engineered to auto add flags for bit fields in the User object.  Should there be a comment in both places minimally?  This is, at most, a nit.

jimabramson replied:
I have not gone ahead and added this, because I'd rather make it in a subsequent change and it seems a little YAGNI.  superuser and staff are the standard Django permission bits.

e0d commented:
General comment, why no delete commands?  For example, the ansible idiom for managing these sorts to entities would typically include a state field where value would be something like present or absent.  If we want to drive from config there's a point were we need to acknowledge the existence of an entity in order to specify that we would like it to be removed.  The alternative would  be to handle deletes manually.  Perhaps the thought is that the builtin Django commands are sufficient for deleting users?

jimabramson commented:
@e0d  I've addressed most of your comments, renamed the commands, and implemented a `--remove` option for both.   A few things that don't make sense to do in this PR but can be done in subsequent ones.

Are you OK with renaming this repo/package edx-management-commands?

e0d commented:
@jimabramson I'm fine with renaming.  Changes LGTM

clintonb commented on .coveragerc:
```diff
@@ -0,0 +1,13 @@
+omit =
+    edx_management_commands/settings*
+    edx_management_commands/conf*
+    edx_management_commands/management_commands/urls.py
```
Why are these two files omitted?

jimabramson replied:
take a look at the urls.py and views.py files, whose docstrings should explain why.

clintonb commented on .travis.yml:
```diff
@@ -0,0 +1,21 @@
+language: python
+python:
+    - "2.7"
```
Would you add support for Python 3.5 so we can use this for newer IDAs (e.g. course-discovery)?

peter-fogg replied:
@clintonb I've Python 3 to the tests, but it seems like Django gives different error messages depending on which version it's running. I suspect it has to do with optparse (old) versus argparse (new), but I haven't tracked it down just yet.

peter-fogg replied:
See https://travis-ci.org/edx/edx-django-extensions/builds/108105435.

peter-fogg replied:
Worked around by [checking the Python version](https://github.com/edx/edx-django-extensions/pull/1/files#diff-30e3fbe61c313a2576e8c69bd2eca096R76). I don't feel great about branching on the language version, but the error message is helpful in either case ("too few arguments" in 2.7 versus "the following arguments are required: group_name" in 3.5) and it seems useful to assert that the error messages are correct

clintonb commented on .travis.yml:
```diff
@@ -0,0 +1,21 @@
+    - pip install -r requirements/test.txt
+script:
+    - make validate
+branches:
```
I have yet to find good documentation regarding why this is necessary. I personally like the idea of testing the branch as well as the branch merged into master, which is what happens when you remove this directive.

clintonb commented on .tx/config:
```diff
@@ -0,0 +1,14 @@
+host = https://www.transifex.com
+
+[edx-platform.edx-django-extensions]
+file_filter = edx-django-extensions/conf/locale/<lang>/LC_MESSAGES/django.po
```
Remember to actually create this project at Transifex.

jimabramson replied:
looks like i missed the repo rename here.   will fix.

clintonb commented on AUTHORS:
```diff
@@ -0,0 +1 @@
+Jim Abramson <jsa@edx.org>
```
Do we need `CONTRIBUTORS.txt` _and_ this file?

jimabramson replied:
uh, no.   I'll move this to CONTRIBUTORS.txt

clintonb commented on Makefile:
```diff
@@ -0,0 +1,67 @@
+local-requirements:
+	pip install -qr requirements/local.txt --exists-action w
+
+requirements:
```
Can we switch this such that requirements installs local requirements, and use a separate target for production requirements? It's a small increase in convenience and efficiency improvement for developers.

cpennington replied:
I'm not convinced that `make requirements` should be local and `make prod-requirements` should be production rather than `make local-requirements` being local. Although, looking at the help text above, you should either fix the docs or fix the code.

peter-fogg replied:
I personally prefer `make requirements` for local development. `make production-requirements` usually will get run as part of the automated build process, whereas `make local-requirements` might get typed multiple times per developer-day. Why not Huffman code our command line?

clintonb commented on README.rst:
```diff
@@ -1 +1,60 @@
+
+EdX Management Commands  |Travis|_ |Codecov|_
+===========================================
+.. |Travis| image:: https://travis-ci.org/edx/edx-management-commands.svg?branch=master
```
These paths are incorrect.

clintonb commented on README.rst:
```diff
@@ -1 +1,60 @@
+.. |Codecov| image:: http://codecov.io/github/edx/edx-management-commands/coverage.svg?branch=master
+.. _Codecov: http://codecov.io/github/edx/edx-management-commands?branch=master
+
+This repository is intended for Django 1.9 applications and utilities for reuse within
```
1.9, or 1.8+?

clintonb commented on README.rst:
```diff
@@ -1 +1,60 @@
+Documentation
+-------------
+
+The docs for EdX Management Commands are on Read the Docs:  https://edx-management-commands.readthedocs.org.
```
This path may be invalid.

jimabramson replied:
came from cookiecutter - should it be yanked until that's online?

clintonb replied:
This is fine. I made the original comment before your clarification of the repo name?

clintonb commented on docs/commands.rst:
```diff
@@ -0,0 +1,91 @@
+
+manage_group
+------------
+The ``manage_group`` command is used to ensure that a specific group exists, and that it has a specific set of
```
Minor: Link to the Group model docs at https://docs.djangoproject.com/en/1.8/ref/contrib/auth/#group-model.

cpennington replied:
You can probably use `intersphinx` to have sphinx figure out how to make the correct doc links.

clintonb commented on docs/commands.rst:
```diff
@@ -0,0 +1,91 @@
+The above will ensure that ``groucho`` will be assigned to ``mygroup`` and ``privileged_group``, and removed from any
+others to which they may already have been assigned.
+
+Note that if any of the groups specified with ``-g`` do not exist, the command will output a warning, but continue.
```
Clarify that the group is not created.

clintonb commented on docs/commands.rst:
```diff
@@ -0,0 +1,91 @@
+
+    $ python manage.py manage_user groucho groucho@example.com --remove
+
+In the above example, the matching user would be deleted if it existed, and the command would do nothing otherwise.
```
This sentence is grammatically incorrect.

peter-fogg replied:
I think it's fine... If the user exists, it gets deleted. If it does not, the command does nothing.

clintonb commented on docs/conf.py:
```diff
@@ -0,0 +1,256 @@
+# html_file_suffix = None
+
+# Output file base name for HTML help builder.
+htmlhelp_basename = 'edx-management-commandsdoc'
```
Replace edx-management-commands with edx-django-extensions throughout the doc.

clintonb commented on edx_management_commands/management_commands/management/commands/manage_group.py:
```diff
@@ -0,0 +1,131 @@
+"""
```
Now I'm just confused. Should this package be called `edx_django_extensions`?

jimabramson replied:
@clintonb the original name for this repo / project was edx-django-extensions.   over the course of review / feedback it's been renamed to edx-management-commands.  sorry for confusion.  i think i've updated everything necessary other than the name of the repo itself.

clintonb commented on edx_management_commands/management_commands/management/commands/manage_user.py:
```diff
@@ -0,0 +1,131 @@
+from django.utils.translation import gettext as _
+
+
+@transaction.atomic
```
Should this be on `handle()`?

clintonb commented on edx_management_commands/management_commands/tests.py:
```diff
@@ -0,0 +1,307 @@
+        self.set_group_permissions(initial_group_permissions)
+
+        # no group name
+        with self.assertRaises(CommandError) as exc_context:
```
Can each of these groups be split into separate tests?

clintonb commented on edx_management_commands/management_commands/urls.py:
```diff
@@ -0,0 +1,15 @@
+"""
```
Most external Django apps I see use a separate test project that has settings and urls.py. That way you don't need to add unnecessary files.

clintonb commented on edx_management_commands/management_commands/urls.py:
```diff
@@ -0,0 +1,15 @@
+from . import views
+
+urlpatterns = [
+    url(r'^$', views.index, name='index'),
```
Can you just use an empty array?

clintonb commented on setup.py:
```diff
@@ -0,0 +1,31 @@
+    ],
+    packages=find_packages(exclude=['tests']),
+    install_requires=[
+        'django<=1.8.7',
```
`django>=1.8.7,<1.9`

clintonb commented on setup.py:
```diff
@@ -0,0 +1,31 @@
+    tests_require=[
+        'ddt',
+        'edx-lint',
+        'coverage==3.7.1',
```
`coverage>=3.7.1`

It might also be useful to start with 4.x.

nedbat replied:
If we aren't using "setup.py test" (are we?) then I'm not sure there's a point to `tests_require`.

peter-fogg replied:
Since we're not using `setup.py test` I'm going to remove the `tests_require`.

clintonb commented on pylintrc_tweaks:
```diff
@@ -0,0 +1,8 @@
+const-rgx = (([A-Z_][A-Z0-9_]*)|(__.*__)|log|urlpatterns|logger|User)$
+
+[MESSAGES CONTROL]
+DISABLE+= ,invalid-name
```
Add a new line.

peter-fogg replied:
Looks like this is actually something that should be handled in the cookiecutter repo: https://github.com/edx/cookiecutter-django-ida/commit/d25ca33bef48344ee01786e1f1a6f35c815e8c30#diff-4eddf84cd665e6a6697c51ed3429907fR8

peter-fogg replied:
https://github.com/edx/cookiecutter-django-ida/pull/18

clintonb commented on .travis.yml:
```diff
@@ -0,0 +1,21 @@
+    - "sh -e /etc/init.d/xvfb start"
+install:
+    - pip install -U pip wheel codecov
+    - pip install -r requirements/test.txt
```
This file does not exist.

clintonb commented on requirements.txt:
```diff
@@ -0,0 +1,7 @@
+# NOTE this requirements file is for tests.
+# Runtime requirements are handled in setup.py
+.
```
`-e .`

clintonb commented on setup.py:
```diff
@@ -0,0 +1,31 @@
+    version='0.0.1',
+    description='Reusable automation and configuration utilities for edX Django applications',
+    author='edX',
+    url='https://github.com/edx/edx-management-commands',
```
Incorrect URL.

jimabramson commented on edx_management_commands/settings.py:
```diff
@@ -0,0 +1,26 @@
+"""
+
+SECRET_KEY = 'insecure-secret-key'
+ROOT_URLCONF = 'edx_user_tools.urls'
```
i will fix this too even though it is borderline useless

cpennington commented on LICENSE.txt:
```diff
@@ -0,0 +1,671 @@
+GNU AFFERO GENERAL PUBLIC LICENSE
```
Since this isn't a core part of the platform, should it be Apache instead of AGPL?

peter-fogg replied:
@nedbat, what's your view on this? I'm not totally clear on our licensing policy for libraries.

nedbat commented on requirements.txt:
```diff
@@ -0,0 +1,7 @@
+# Runtime requirements are handled in setup.py
+.
+ddt
+edx-lint
```
You should pin the version you use.

nedbat commented on requirements.txt:
```diff
@@ -0,0 +1,7 @@
+.
+ddt
+edx-lint
+coverage==3.7.1
```
Since we're pinning, might as well use the latest: 4.0.3

nedbat replied:
Or, a range as Clinton suggests elsewhere.

peter-fogg commented on edx_management_commands/management_commands/management/commands/manage_group.py:
```diff
@@ -0,0 +1,131 @@
+            self.stderr.write(_('Removed group: "{}"').format(group_name))
+        except Group.DoesNotExist:
+            self.stderr.write(_('Did not find a group with name "{}" - skipping.').format(group_name))
+            return
```
`return` isn't really necessary here.

peter-fogg commented on edx_management_commands/management_commands/management/commands/manage_group.py:
```diff
@@ -0,0 +1,131 @@
+        old_permissions, new_permissions = [], []
+        group, created = Group.objects.get_or_create(name=group_name)  # pylint: disable=no-member
+
+        if created:
```
Is there a reason that remove is handled by a separate method, but the other logic isn't split out? Might be nice to break up this 100-line-ish method.

peter-fogg commented on edx_management_commands/management_commands/management/commands/manage_user.py:
```diff
@@ -0,0 +1,131 @@
+            self.stderr.write(_('Did not find a user with username "{}" - skipping.').format(username))
+            return
+
+        user = users[0]
```
Just to avoid this, why not `try: user = get_user_model().objects.get(username=username)` and catch `DoesNotExist`?

peter-fogg commented on edx_management_commands/management_commands/management/commands/manage_user.py:
```diff
@@ -0,0 +1,131 @@
+        old_groups, new_groups = [], []
+        user, created = get_user_model().objects.get_or_create(username=username)
+
+        if created:
```
Wondering the same thing here about why `_handle_remove` is broken out for testing but the other parts aren't.

peter-fogg commented on edx_management_commands/management_commands/management/commands/manage_user.py:
```diff
@@ -0,0 +1,131 @@
+                )
+                user.groups.remove(rg)
+        else:
+            self.stderr.write(_('Not removing user "{}" from any groups').format(username))
```
For bonus Python points you could use `for`/`else`: https://docs.python.org/2/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops

peter-fogg replied:
Oh, disregard that, I read the above loop wrong.

peter-fogg commented on edx_management_commands/management_commands/tests.py:
```diff
@@ -0,0 +1,307 @@
+        """
+        DRY helper.
+        """
+        self.assertEqual(set(group_names), set([g.name for g in Group.objects.all()]))  # pylint: disable=no-member
```
Could use a set comprehension in the second argument.

peter-fogg commented on edx_management_commands/management_commands/tests.py:
```diff
@@ -0,0 +1,307 @@
+        """
+        self.assertEqual(
+            set(permission_codenames),
+            set([p.codename for p in Group.objects.get(name=group_name).permissions.all()])  # pylint: disable=no-member
```
Same note about set comprehensions.

peter-fogg commented on edx_management_commands/management_commands/tests.py:
```diff
@@ -0,0 +1,307 @@
+
+    def set_group_permissions(self, group_permissions):
+        """
+        Sets up a before-state for groups and permissions in tests, which can checked
```
"which can be checked"

Commit:
Fixes after taking over from Jim.

Split up management logic, refactored adding/removing
groups/permissions a bit, added Python 3 compatibility, and fixed up
testing in Travis.

peter-fogg commented:
@cpennington @clintonb @nedbat Could whichever of you are currently reviewing this go ahead and take another look? I've made a few changes and addressed various bits of feedback. I have a few questions about supporting Python 3 and the project structure which I'll ask about tomorrow.

adampalay commented on edx_management_commands/management_commands/management/commands/manage_user.py:
```diff
@@ -0,0 +1,135 @@
+            return
+        self._check_email_match(user, email)
+        self.stderr.write(_('Removing user: "{}"').format(user))
+        user.delete()
```
We need to make sure we manage expectations about what users are and what can happen to them as part of this

cpennington commented on edx_management_commands/management_commands/management/commands/manage_group.py:
```diff
@@ -0,0 +1,140 @@
+                # name.max_length won't be enforced by the db.
+                # See also http://www.sqlite.org/faq.html#q9
+                group.full_clean()
+            except ValidationError, exc:
```
This should be `except ValidationError as exc`.

cpennington commented on edx_management_commands/management_commands/management/commands/manage_group.py:
```diff
@@ -0,0 +1,140 @@
+        new_permissions = self._resolve_permissions(permissions or [])
+
+        add_permissions = [p for p in new_permissions if p not in old_permissions]
+        remove_permissions = [p for p in old_permissions if p not in new_permissions]
```
Why do we explicitly specify each permission to add/remove, rather than just setting `group.permissions.set(permissions)`, which does it all for us?

peter-fogg replied:
I can't speak to Jim's thought process here, but it seems valuable to have explicit logging of which permissions are being added and removed. It does make sense to directly set the permissions, though, so I'll rework this a little.

cpennington commented on edx_management_commands/management_commands/management/commands/manage_user.py:
```diff
@@ -0,0 +1,135 @@
+                    attribute=attribute, username=user.username, new_value=new_value
+                )
+            )
+            setattr(user, attribute, new_value)
```
Why not just set the value explicitly? Is this just so that we can log the changes that were made?

peter-fogg replied:
I like having this here to explicitly log the changes.

cpennington commented on edx_management_commands/management_commands/management/commands/manage_user.py:
```diff
@@ -0,0 +1,135 @@
+                self.stderr.write(_('Could not find a group named "{}" - skipping.').format(group_name))
+
+        add_groups = [p for p in new_groups if p not in old_groups]
+        remove_groups = [p for p in old_groups if p not in new_groups]
```
Again, maybe just use `user.groups.set(...)`.

cpennington replied:
If you need to log the changes, you could just log `set(new_groups) - set(old_groups)` and `set(old_groups) - set(new_groups)`.

peter-fogg commented:
@clintonb @cpennington How's this looking now? I've been keeping my commits small and separate for ease of reviewing.

cpennington commented:
One minor comment on set manipulation, but once that's addressed, :+1:
